### PR TITLE
[VLBOCKS-2885] Feat: opt in callmessage

### PR DESCRIPTION
**Pull request by kpchoy**
**Contributing to Twilio**

> All third party contributors acknowledge that any contributions they provide will be made under the same open source license that the open source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### JIRA link(s):

- [VBLOCKS-2885](https://twilio-engineering.atlassian.net/browse/VBLOCKS-2885)

### Description

Opt in feature for call message types

Manual testing:
- WS Invite has callMessageEvents
- WS Answer has callMessageEvents

### Before review
* [x] Updated CHANGELOG.md if necessary
* [x] Added unit tests if necessary
* [x] Updated affected documentation
* [x] Verified locally with `npm test`
* [x] Manually sanity tested running locally
* [x] Ready for review

### Before merge
* [x] Got one or more +1s
* [x] Squashed erroneous commits if necessary
* [ ] Re-tested if necessary

### CODE REVIEW

**Review by charliesantos — CHANGES_REQUESTED** [review]
@kpchoy this is for eventing 2.0. Please submit to feature/eventing_2.0 branch instead.

File: lib/twilio/device.ts [hunk 1]
```diff
@@ -342,6 +342,7 @@ class Device extends EventEmitter {
    */
   private readonly _defaultOptions: IExtendedDeviceOptions = {
     allowIncomingWhileBusy: false,
+    callMessageEvents: ['user-defined-message'],
```
Comment (charliesantos) [inline lib/twilio/device.ts:345]: Default should be an empty array. The auto-subscription happens in the backend.

File: lib/twilio/device.ts [hunk 2]
```diff
@@ -1041,6 +1043,7 @@ class Device extends EventEmitter {
         this._activeCall.disconnect();
         this._removeCall(this._activeCall);
       },
+      callMessageEvents: this._options.callMessageEvents,
```
Comment (charliesantos) [inline lib/twilio/device.ts:1046]: What happens to this array inside the Call object after passing it from here?

**Review by charliesantos — CHANGES_REQUESTED** [review]
@kpchoy the implementation seems to be incomplete. Is this intentional, or did we forget to commit the rest of the changes? In this PR, we're only passing the callMessageEvents from device to the call object. The wiring from call object all the way to sending to the server is missing. Please check the whole section of the blueprint that's linked in the JIRA.

File: lib/twilio/device.ts [hunk 3]
```diff
@@ -1840,6 +1843,12 @@ namespace Device {
      */
     appVersion?: string;
 
+    /**
+     * The callMessage types a client subscribes to. 'user-defined-message' will always be opted in by
+     * default.
+     */
```
Comment (charliesantos) [inline lib/twilio/device.ts:1849]: >An array of [call message types](https://twilio.github.io/twilio-voice.js/interfaces/voice.call.message.html#messagetype) to subscribe to. Currently, only 'user-defined-message' is supported. More message types will be added in the future. See [call resource](https://www.twilio.com/docs/voice/api/call-resource) documentation for more details.
>
>When subscribed, the SDK will emit the message via [messageReceivedEvent](https://twilio.github.io/twilio-voice.js/classes/voice.call.html#messagereceivedevent).

Comment (charliesantos) [reply lib/twilio/device.ts:1849]: I used markdown style for the links. Please update the links for call message types and messageReceivedEvent to use typdoc/tsdoc style.

File: lib/twilio/rtc/peerconnection.ts [hunk 4]
```diff
@@ -922,9 +922,9 @@ PeerConnection.prototype.makeOutgoingCall = function(params, signalingReconnectT
   function onOfferSuccess() {
     if (self.status !== 'closed') {
       if (signalingReconnectToken) {
-        self.pstream.reconnect(self.version.getSDP(), self.callSid, signalingReconnectToken);
+        self.pstream.reconnect(self.version.getSDP(), self.callSid, signalingReconnectToken, callMessageEvents);
```
Comment (charliesantos) [inline lib/twilio/rtc/peerconnection.ts:925]: pstream.reconnect() is also used in `Call.ts`. Please update that part.

**Review by charliesantos — CHANGES_REQUESTED** [review]
Left some comments. Also, please add tests for call.ts and device.ts changes.

File: lib/twilio/device.ts [hunk 5]
```diff
@@ -1844,8 +1844,11 @@ namespace Device {
     appVersion?: string;
 
     /**
-     * The callMessage types a client subscribes to. 'user-defined-message' will always be opted in by
-     * default.
+     * An array of {@link Call.Message.messageType} to subscribe to. Currently, only 'user-defined-message'
+     * is supported. More message types will be added in the future.
+     * See [call resource](https://www.twilio.com/docs/voice/api/call-resource) documentation for more details.
+     *
+     * When subscribed, the SDK will emit the message via {@link Call._onMessageReceived}.
```
Comment (charliesantos) [inline lib/twilio/device.ts:1851]: We should link the public one like below. You can also test the docs by running `npm run docs:ts` then open `docs/index.html` from your browser.
```suggestion
     * When subscribed, the SDK will emit the message via {@link Call.messageReceivedEvent}.
```

File: lib/twilio/call.ts [hunk 6]
```diff
@@ -1228,6 +1230,7 @@ class Call extends EventEmitter {
         this._mediaHandler.version.getSDP(),
         this.parameters.CallSid,
         this._signalingReconnectToken,
+        this._options.callMessageEvents,
```
Comment (charliesantos) [inline lib/twilio/call.ts:1233]: Let's add unit test to this as well. You can pass callMessageEvents [here](https://github.com/twilio/twilio-voice.js/blob/VBLOCKS-2885/opt-in-call-message/tests/unit/call.ts#L1914). Then perform your assertion [here](https://github.com/twilio/twilio-voice.js/blob/VBLOCKS-2885/opt-in-call-message/tests/unit/call.ts#L1931).

**Review by charliesantos — CHANGES_REQUESTED** [review]
Super close Kevin! I left a couple more comments.

File: lib/twilio/device.ts (hunk 5 above)
Comment (kpchoy) [reply lib/twilio/device.ts]: oops, forgot to check!

**Review by charliesantos — APPROVED** [review]
:shipit:

**Pull request merged**